### ISSUE TITLE
Open-ended page ranges

### ISSUE THREAD

**Issue by denismaier** [citation-style-language/schema#372]
There's been a [request on pandoc-discuss](https://groups.google.com/g/pandoc-discuss/c/aPl79Of7MZ4) regarding open-ended page ranges:

> Is it possible to have pandoc-citeproc interpret a locator like 12ff as multiple pages? Currently, I get p. 12ff while pp. 12ff would be correct. 

I don't think there's support for that at the moment, right? Also, I don't think 1.1 will add support for this yet, correct? Or, am I just missing something here?

Arguably, open-ended page-ranges are not really useful for readers, but in some traditions, it's rather common. In German language academia you'll often even find that page-ranges are always shortened to starting page + "ff.". So, "215-243" will be "215ff." (When writing in German, it's actually not a big deal since the abbreviation for "Seite" and "Seiten" is in both cases "S.").

Both citeproc-js and pandoc-citeproc seem to have some support for this already. A locator "215-" will be interpreted as an open-ended range, and the label will be adjusted accordingly, so in citeproc-js the result is "pp. 215-", with pandoc-citeproc it's "pp. 215".

I suggest we should add some support for this. Some possibilities:

- a new term `psqq` à la biblatex that could be used with open-ended page-ranges.
- a new term `psq`, and a new option to support "12-13"=>"12f." like behavior.
- a new option to support unconditional collapsing of page-ranges into first page + "ff.": "12-25" => "12ff."

Thought? @jgm @cormacrelf @fbennett @andras-simonyi @bdarcus @bwiernik @georgd

**Comment by bdarcus** [comment]
In general, 1.1 will have support for open-ended ranges on data input (not
sure we've explicitly dealt with styling yet); locators and dates.

**Comment by denismaier** [comment]
Yeah, input should be safe. Question is just what should happen in styles.

**Comment by bwiernik** [comment]
Allow me first to complain that this is one of those archaic style conventions like op cit that I wish would go away. 😜

It's not really clear to me exactly how open-ended ranges addresses the data needs for `183f.` versus `183ff.` It seems almost to be unrelated. 

For example, if a cited range were `183–184`, shouldn't that be collapsed to `183f.`? Similarly, if the range is `183–187`, should that be collapsed to `183ff.`? What is the number of pages where the format changes from `ff.` to actually spelling out the range? 

It seems more like the `183f.` format would need to be addressed using a new page range format option alongside minimal, chicago, etc. 

If I'm misunderstanding, could you elaborate on when the `183ff.` format versus an enumerated range is used?

(With respect to what is the expected output for open-ended ranges in typical styles? Would it be something like `pp. 183–` ?)

**Comment by georgd** [comment]
> Allow me first to complain that this is one of those archaic style conventions like op cit that I wish would go away. 😜

I agree. But in some areas this convention is as firmly established as anything. So I don’t see it going away.

> It's not really clear to me exactly how open-ended ranges addresses the data needs for `183f.` versus `183ff.` It seems almost to be unrelated.

Agreed, the two are unrelated in respect to the question of open-ended ranges. The former is not open-ended but simply a two-page range. They are only related with respect to the proposal by @denismaier that a new term should be added which would be singular for the two-page range and plural for any other page range.

> For example, if a cited range were `183–184`, shouldn't that be collapsed to `183f.`? Similarly, if the range is `183–187`, should that be collapsed to `183ff.`? What is the number of pages where the format changes from `ff.` to actually spelling out the range?

All the styles I know of to require page-range collapsing (all of them German language, in different areas, mostly history, business, and law), do so unconditionally. In such a style, a two-page range is always 183 f., and any other page-range is always 183 ff.

**Comment by bwiernik** [comment]
Seems like adding a new page range option collapse option would be the approach then.

**Comment by denismaier** [comment]
`183f.` is indeed not open-ended. It's a two-page range, with special collapsing behavior being applied. So, yes, a new  new page range format (together with an appropriate term) would be the way to go.

Regarding `ff.` there are two scenarios:

1. A style might require page-range collapsing unconditionally, as elaborated by @georgd. So, `183-185` will always be rendered as `183ff.`. We don't actually have an open-ended range here. So that would be just another new page range format.
2. A user might want, for whatever reason, to occassionally use proper open-ended ranges like `183-`, with the desired output `183ff.`, or maybe `183-`, which would correspond to Chicago's open ended year range format. I don't know of a style that uses `pp. 183-` though. Chicago's take on this, 14.149:
> Only when referring to a section for which no final number can usefully be given should _ff._ (“and the following pages, paragraphs, etc.”) be resorted to.

So, it seems this could boil down to this:

- one term `psq` with two forms:
  - singular: `f.`
  - plural: `ff.`
  (Alternatively, the plural could be a term `psqq` on its own.)
- new page range options:
  - `psq` => collapse two-page ranges into "starting page" + "f."
  - `psqq` => same as `psq` + collapse multi-page-ranges into "starting page" + "ff."
- Add a new attribute `@use-psqq-with-open-ended-page-range` (the naming might be improved ;-)); if true, open-ended page ranges are rendered as "starting-page" + "ff."

**Comment by bdarcus** [comment]
> 2\. A user might want, for whatever reason, to occassionally use proper open-ended ranges ...

If you mean, for some locators use ff, but in others not, then not clear to me we should support this requirement.

**Comment by denismaier** [comment]
> If you mean, for some locators use ff, but in others not, then not clear to me we should support this requirement.

Yes, I mean this:
If a user enters a page-range like `183-187`, format it according to the chosen page range format.
But, if they enter `183-`, format it as `183ff.`

We might not even need a new option for this. If a user really wants an open-ended range, I don't see how that should be formatted differently. (The `ff.` must be localized of course.)

**Comment by bwiernik** [comment]
What's psq mean?

**Comment by denismaier** [comment]
psq is borrowed from biblatex; it's a combination of two abbreviations: p + sq => the following page.

**Comment by bwiernik** [comment]
What's the sq stand for?

**Comment by denismaier** [comment]
sq = lat. sequens; the following one
see https://www.collinsdictionary.com/dictionary/english/sq
sqq is the plural

We might as well decide to not use latin abbreviations if there are suggestions for a different name.

**Comment by bwiernik** [comment]
`following-page`?

**Comment by denismaier** [comment]
> `following-page`?

No please. That would be way to obvious.

**Comment by denismaier** [comment]
I've updated my suggestion [above](https://github.com/citation-style-language/schema/issues/372#issuecomment-686758866).

**Comment by bdarcus** [comment]
> * Add a new attribute ~`@use-psqq-with-open-ended-page-range`~ `@use-following-pages-with-open-ended-page-range`  (the naming might still be improved ;-)); if true, open-ended page ranges are rendered as "starting-page" + "ff."

Rather than a boolean with a specific name, how about a more general name with a specific value; say `@page-open-range-format="following"`

**Comment by georgd** [comment]
> * new page range options:
>   
>   * ~`psq`~  `collapse-following-page` => collapse two-page ranges into "starting page" + "f."
>   * ~`psqq`~  `collapse-following-pages`=> same as ~`psq`~  `following-page` + collapse multi-page-ranges into "starting page" + "ff."
> * Add a new attribute ~`@use-psqq-with-open-ended-page-range`~ `@use-following-pages-with-open-ended-page-range`  (the naming might still be improved ;-)); if true, open-ended page ranges are rendered as "starting-page" + "ff."

@denismaier I understand that with this setup you suggest that only in certain cases, an open-ended page-range will be rendered with "starting-page" + "ff.". IMO, it should be the other way around. When `collapse-following-page` is set, open-ended page-ranges always should be rendered with "ff." (to me, "S. 123 ff." looks less confusing than "S. 123-"). At least, I’ve never seen such open-ended range with anything but "ff.".

**Comment by denismaier** [comment]
Maybe we can simplify the terms even more: 
we could just call it `following`. Then, this term could also be used for open-ended year ranges. We can then have two attributes: `@page-open-range-format="following"`. 
`@year-open-range-format="following"`.

**Comment by denismaier** [comment]
> @denismaier I understand that with this setup you suggest that only in certain cases, an open-ended page-range will be rendered with "starting-page" + "ff.". IMO, it should be the other way around. When `collapse-following-page` is set, open-ended page-ranges always should be rendered with "ff." (to me, "S. 123 ff." looks less confusing than "S. 123-"). At least, I’ve never seen such open-ended range with anything but "ff." [edit] in styles that use this kind of collapsing.

Good point. OTOH, you might want to use this kind of formatting **only** if the user supplies an open-ended range, but you might not want to collapse "183-187".

**Comment by georgd** [comment]
> > @denismaier I understand that with this setup you suggest that only in certain cases, an open-ended page-range will be rendered with "starting-page" + "ff.". IMO, it should be the other way around. When `collapse-following-page` is set, open-ended page-ranges always should be rendered with "ff." (to me, "S. 123 ff." looks less confusing than "S. 123-"). At least, I’ve never seen such open-ended range with anything but "ff." [edit] in styles that use this kind of collapsing.
> 
> Good point. OTOH, you might want to use this kind of formatting **only** if the user supplies an open-ended range, but you might not want to collapse "183-187".

That portion of my answer got lost. I wanted to state that the closed page-range with more than two pages might be seen as the "marked" case which would not be collapsed in certain circumstances.

> Perhaps `collapse-following-pages` could automatically set `@page-open-range-format="following"`. That would be what you describe. But you could still use `@page-open-range-format="following"` with other page formats as well.

I think, `collapse-following-page` and `collapse-following-pages` both should implicitly set `@page-open-range-format="following"` (and perhaps `collapse-following-pages` should set `collapse-following-page` as well?)

**Comment by denismaier** [comment]
> (and perhaps `collapse-following-pages` should set `collapse-following-page` as well?)

Sure. These are not booleans. Theye are selected as `page-range-format="collapse-following-page"` or `page-range-format="collapse-following-pages"`. 

`page-range-format="collapse-following-pages"` would collapse two-page ranges to "starting page" + "f."; longer finite page ranges will trigger the "ff." collapsing.

**Comment by georgd** [comment]
> > (and perhaps `collapse-following-pages` should set `collapse-following-page` as well?)
> 
> Sure. These are not booleans. They are selected as `page-range-format="collapse-following-page"` or `page-range-format="collapse-following-pages"`.
> 
> `page-range-format="collapse-following-pages"` would collapse two-page ranges to "starting page" + "f."; longer finite page ranges will trigger the "ff." collapsing.

I see. Well, I think that’s a good setting.

**Comment by bwiernik** [comment]
I'm wondering if we really need two mechanisms here (`page-range-format` and `page-open-range-format`). If we added `page-range-format="collapse-following-pages"` (which would use `f.` or `ff.` as appropriate) and added the open-range formatting specification to `page-range-format="chicago"`, would that suffice?

It seems like the `18ff.` style is very uncommon, so I'm not sure we need to provide separate control of the open page range formatting.
